src/Microsoft.AspNetCore.Routing/Matchers/TreeMatcher.cs

```diff
@@ -141,37 +141,31 @@ internal class TreeMatcher : Matcher
 
         private Task SelectEndpointAsync(HttpContext httpContext, IEndpointFeature feature, IReadOnlyList<MatcherEndpoint> endpoints)
         {
-            var bestEndpoint = _endpointSelector.SelectBestCandidate(httpContext, endpoints);
+            MatcherEndpoint endpoint;
 
-            // REVIEW: Note that this code doesn't do anything significant now. This will eventually incorporate something like IActionConstraint
-            switch (endpoints.Count)
+            try
             {
-                case 0:
-                    {
-                        Log.MatchFailed(_logger, httpContext);
-                        return Task.CompletedTask;
-                    }
-
-                case 1:
-                    {
-                        var endpoint = endpoints[0];
-                        Log.MatchSuccess(_logger, httpContext, endpoint);
-
-                        feature.Endpoint = endpoint;
-                        feature.Invoker = endpoint.Invoker;
+                endpoint = (MatcherEndpoint)_endpointSelector.SelectBestCandidate(httpContext, endpoints);
+            }
+            catch (Exception ex)
+            {
+                // Handle AmbiguousMatchException from EndpointSelector
+                return Task.FromException(ex);
```

Comment:
uhhhhhh - this shouldn't be super needed.

Reply:
Note that this also defeats teh logging

Reply:
As far as I know it is best practice to return a task with an exception instead of throwing from methods that return task.

Logging happens in the EndpointSelector before the AmbiguousMatchException is thrown.
The await on MatchAsync will throw the exception from the task.

Reply:
Why though? There are lots of best practices... who's to say which best practices are best?

Much folk wisdom related to `Task` dates back to .NET 4.0 (pre-await) and isn't relevant anymore.

Reply:
It's to do with getting a `Task` from a method without awaiting it. Since the method isn't public and we're the only callers I guess it doesn't matter.

Reply:
That whole method doesn't need to be async anyway. I'll refactor it.